tracks/python/exercises/book-store/mentoring.md

```diff
@@ -0,0 +1,301 @@
+    be raised. While perhaps this is unrealistic input, it might be an interesting point to discuss,
+    as this sort of thing can be a problem in practice.
+
+
+The iterative approach mitigates these concerns but can produce somewhat less elegant code. A
+straightforward approach is to use a `for` loop for each of the five titles:
+
+```python
+def total(basket):
+    table = {(0, 0, 0, 0, 0): 0}
+    target_quantities = (n1, n2, n3, n4, n5) = count_quantities(basket)
+    for i1 in range(n1+1):
+        for i2 in range(n2+1):
+            for i3 in range(n3+1):
+                for i4 in range(n4+1):
+                    for i5 in range(n5+1):
+                        quantities = (i1, i2, i3, i4, i5)
+                        if sum(quantities) == 0:
+                            continue
+                        table[quantities] = min(
+                            combo_price(combo) 
+                            + table[remove(quantities, combo)]
+                            for combo_size in DISCOUNTS
+                            for combo in combinations(TITLES, combo_size)
+                            if have_enough_for(quantities, combo)
+                        )
+    return table[target_quantities]
+```
+
+This works, but:
+
+-   It's ugly;
+
+-   It's not very extensible - what if we wanted to increase number of titles?  There's a general
+    principle to be learnt: try to avoid hard-coding data into the program.
+
+
+An alternative is to recur w.r.t. the different items:
+
+```python
+def total(basket):
+    targets = count_quantities(basket)
+    table   = build_table(targets)
+    return table[targets]
+
+def build_table(targets, quantities=(), remaining=len(TITLES), table=None):
+    table = table or {(0,) * remaining: 0}
+    if remaining:
+        target = targets[remaining-1]
+        for quantity in range(target+1):
+            build_table(targets, (quantity,)+quantities, remaining-1, table)
+    elif sum(quantities) > 0:
+        table[quantities] = min(
+            combo_price(combo) + table[remove(quantities, combo)]
+            for combo_size in DISCOUNTS
+            for combo in combinations(TITLES, combo_size)
+            if have_enough_for(quantities, combo)
+        )
+    return table
+```
+
+Here are some thoughts on this implementation:
+
+-   This might be a good compromise. It is recursive, but along a different axis and wont recur too
+    deeply (unless list of titles is changed to be very large; but cross that bridge if you ever
+    come to it).
+
+-   It's important to understand mutability to interpret how this implementation works --- this can
+    be a stumbling block for beginners.  How is the data shared? Only one table is ever created, so
+    everyone is putting their data into that same table.
+
+-   As with the recursive approach, this can be optimized with a canonical form.
+
+-   Why use dict of tuples and not a 5-dimensional array? The latter is complicated; the author 
```

Comment:
Generally I would avoid references to "the author" and "the author's machine" ... this document will, over time, endure many different authorship events.